### PR TITLE
NIFI-10789: set flowfile attributes upon failure/error when fetching …

### PR DESCRIPTION
**Pull request by emiliosetiadarma**
…object in Azure Data Lake Storage

# Summary

[NIFI-10789](https://issues.apache.org/jira/browse/NIFI-10789)
- Write `azure.datalake.storage.*` flowfile attributes when encountering failures to fetch files from Azure Data Lake Storage 

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 8
  - [x] JDK 11
  - [x] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureDataLakeStorage.java [hunk 1]
```diff
@@ -124,14 +134,25 @@ public void onTrigger(ProcessContext context, ProcessSession session) throws Pro
...
+            flowFile = session.putAttribute(flowFile, "azure.datalake.storage.statusCode", String.valueOf(e.getStatusCode()));
+            flowFile = session.putAttribute(flowFile, "azure.datalake.storage.errorCode", e.getErrorCode());
+            flowFile = session.putAttribute(flowFile, "azure.datalake.storage.serviceMessage", e.getServiceMessage());
+            flowFile = session.putAttribute(flowFile, "azure.datalake.storage.errorMessage", e.getMessage());
+            flowFile = session.penalize(flowFile);
+            session.transfer(flowFile, REL_FAILURE);
         } catch (Exception e) {
             getLogger().error("Failure to fetch file from Azure Data Lake Storage", e);
+            // other exception, set a generic failure reason flowfile attribute
+            flowFile = session.putAttribute(flowFile, "azure.datalake.storage.failure.reason", e.getMessage());
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureDataLakeStorage.java:155]: Recommend naming this `azure.datalake.storage.errorMessage` to match the other exception attribute.
```suggestion
            flowFile = session.putAttribute(flowFile, "azure.datalake.storage.errorMessage", e.getMessage());
```

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureDataLakeStorage.java [hunk 2]
```diff
@@ -48,6 +51,13 @@
 @SeeAlso({PutAzureDataLakeStorage.class, DeleteAzureDataLakeStorage.class, ListAzureDataLakeStorage.class})
 @CapabilityDescription("Fetch the provided file from Azure Data Lake Storage")
 @InputRequirement(Requirement.INPUT_REQUIRED)
+@WritesAttributes({
+        @WritesAttribute(attribute = "azure.datalake.storage.statusCode", description = "The HTTP error code (if available) from the failed operation"),
+        @WritesAttribute(attribute = "azure.datalake.storage.errorCode", description = "The Azure Data Lake Storage moniker of the failed operation"),
+        @WritesAttribute(attribute = "azure.datalake.storage.serviceMessage", description = "The Azure Data Lake Storage service message from the failed operation"),
+        @WritesAttribute(attribute = "azure.datalake.storage.errorMessage", description = "The Azure Data Lake Storage error message from the failed operation"),
+        @WritesAttribute(attribute = "azure.datalake.storage.failure.reason", description = "An error message to indicate the cause of the failed operation")
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureDataLakeStorage.java:59]: Recommend removing this attribute and consolidating on the usage of `azure.datalake.storage.errorMessage`:
```suggestion
```

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for the improvement @emiliosetiadarma! The general approach looks good, just one note regarding reuse of the `errorMessage` attribute.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureDataLakeStorage.java (hunk 1 above)
Comment (emiliosetiadarma) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureDataLakeStorage.java:155]: Will make the change!

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureDataLakeStorage.java (hunk 2 above)
Comment (emiliosetiadarma) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureDataLakeStorage.java:59]: Will make the change!

**Review by exceptionfactory — APPROVED** [review]
Thanks for making the adjustments @emiliosetiadarma, the latest version looks good! +1 merging

**Pull request closed without merging**